### PR TITLE
Crac improvements

### PR DESCRIPTION
**Pull request by TheMaskedTurtle**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
- Possibility to add and list contingencies in crac object
- New architecture for thresholds
- Introduction of synchronization concept

### CODE REVIEW

File: data/crac-file/crac-api/src/main/java/com/farao_community/farao/data/crac_api/SynchronizationException.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ *  License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package com.farao_community.farao.data.crac_api;
+
+/**
+ * @author Joris Mancini {@literal <joris.mancini at rte-france.com>}
+ */
+public class SynchronizationException extends Exception {
```
Comment (pengbo-ruidian) [inline data/crac-file/crac-api/src/main/java/com/farao_community/farao/data/crac_api/SynchronizationException.java]: Could we use FaraoException in commons to be consistent with other part of code? or SynchronizationException has something different from FaraoException, then it would be good to note it in the header comment.

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/AbsoluteFlowThreshold.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        this.maxValue = maxValue;
+    }
+
+    @Override
+    public boolean isMinThresholdOvercome(Network network, Cnec cnec) {
+        return false;
+    }
+
+    @Override
+    public boolean isMaxThresholdOvercome(Network network, Cnec cnec) {
```
Comment (pengbo-ruidian) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/AbsoluteFlowThreshold.java]: Here we suppose current limits of Branch.Side.ONE is always equal to Branch.Side.TWO.

File: data/crac-file/crac-io-json/pom.xml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-commons</artifactId>
+            <version>${powsybl.core.version}</version>
+            <type>test-jar</type>
+            <scope>test</scope>
+        </dependency>
+        <dependency> <!--for CracSchemaGeneration-->
+            <groupId>com.fasterxml.jackson.module</groupId>
+            <artifactId>jackson-module-jsonSchema</artifactId>
+            <version>2.3.0</version>
```
Comment (pengbo-ruidian) [inline data/crac-file/crac-io-json/pom.xml]: might be better to add <scope>test</scope> here? or it does not matter at all.

File: data/crac-file/crac-io-json/src/test/java/com/farao_community/farao/data/crac_io_json/JsonImportExportTest.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+        return crac;
+    }
+
+    @Test
+    public void testExportCrac() throws IOException {
+
+        Crac crac = create();
+
+        OutputStream os = new FileOutputStream("file.json");
```
Comment (pengbo-ruidian) [inline data/crac-file/crac-io-json/src/test/java/com/farao_community/farao/data/crac_io_json/JsonImportExportTest.java]: It might be better to not generate a new file.json during the test?
Could use ByteArrayOutputStream() and compare the string?

File: data/crac-file/crac-io-json/pom.xml (hunk 3 above)
Comment (pengbo-ruidian) [reply data/crac-file/crac-io-json/pom.xml]: I mean "<scope>test"

Comment (pengbo-ruidian) [reply data/crac-file/crac-io-json/pom.xml]: I mean \<scope\>test\<\/scope\>

**Comment by baptsegu** [comment]
@TheMaskedTurtle -> review done, except for what is related to Json import/export

File: data/crac-file/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Synchronizable.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+package com.farao_community.farao.data.crac_api;
+
+import com.powsybl.iidm.network.Network;
+
+/**
+ * @author Joris Mancini {@literal <joris.mancini at rte-france.com>}
+ */
+public interface Synchronizable {
+
+    void synchronize(Network network);
```
[the remark this answers is not in the document]
Comment (TheMaskedTurtle) [reply data/crac-file/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Synchronizable.java]: Done !

File: data/crac-file/crac-api/src/main/java/com/farao_community/farao/data/crac_api/SynchronizationException.java (hunk 1 above)
Comment (TheMaskedTurtle) [reply data/crac-file/crac-api/src/main/java/com/farao_community/farao/data/crac_api/SynchronizationException.java]: I added a comment

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/AbsoluteFlowThreshold.java (hunk 2 above)
Comment (TheMaskedTurtle) [reply data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/AbsoluteFlowThreshold.java]: I moved the management of side one and two in getTerminal method in AbstractFlowThreshold

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/RelativeFlowThreshold.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    @Override
+    public boolean isMinThresholdOvercome(Network network, Cnec cnec) {
+        return false;
+    }
+
+    @Override
+    public boolean isMaxThresholdOvercome(Network network, Cnec cnec) throws SynchronizationException {
+        if (Double.isNaN(maxValue)) {
```
[the remark this answers is not in the document]
Comment (TheMaskedTurtle) [reply data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/RelativeFlowThreshold.java]: Instead I assured in the constructor that maxValue is always defined and set to NaN by default

File: data/crac-file/crac-io-json/src/test/java/com/farao_community/farao/data/crac_io_json/JsonImportExportTest.java (hunk 4 above)
Comment (TheMaskedTurtle) [reply data/crac-file/crac-io-json/src/test/java/com/farao_community/farao/data/crac_io_json/JsonImportExportTest.java]: This is more related to Viktor's PR and I think it will be changed in this PR instead

File: data/crac-file/crac-io-json/pom.xml (hunk 3 above)
Comment (TheMaskedTurtle) [reply data/crac-file/crac-io-json/pom.xml]: I changed it

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/AbsoluteFlowThreshold.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        return maxValue < network.getBranch(cnec.getCriticalNetworkElement().getId()).getCurrentLimits(Branch.Side.ONE).getPermanentLimit();
```
[the remark this answers is not in the document]
Comment (TheMaskedTurtle) [reply data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/AbsoluteFlowThreshold.java]: It's adapted now

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/AbsoluteFlowThreshold.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+public class AbsoluteFlowThreshold extends AbstractFlowThreshold {
+
+    public AbsoluteFlowThreshold(Unit unit, Side side, Direction direction, double maxValue) {
+        super(unit, side, direction);
+        this.maxValue = maxValue;
+    }
+
+    @Override
+    public boolean isMinThresholdOvercome(Network network, Cnec cnec) {
+        return false;
```
Comment (terriervik) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/AbsoluteFlowThreshold.java]: Are you going to implement this method differently?

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/RelativeFlowThreshold.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    private double percentageOfMax;
+
+    public RelativeFlowThreshold(Unit unit, Side side, Direction direction, double percentageOfMax) {
+        super(unit, side, direction);
+        this.percentageOfMax = percentageOfMax;
+    }
+
+    @Override
+    public boolean isMinThresholdOvercome(Network network, Cnec cnec) {
+        return false;
```
Comment (terriervik) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/RelativeFlowThreshold.java]: Same as for AbsoluteFlowThreshold

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/VoltageThreshold.java [hunk 10]
```diff
@@ -41,4 +45,24 @@ public double getMaxValue() {
...
+
+    @Override
+    public boolean isMinThresholdOvercome(Network network, Cnec cnec) throws SynchronizationException {
+        return false;
+    }
+
+    @Override
+    public boolean isMaxThresholdOvercome(Network network, Cnec cnec) throws SynchronizationException {
+        return false;
+    }
```
Comment (terriervik) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/VoltageThreshold.java]: Is this finished or TODO?

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/SimpleCrac.java [hunk 11]
```diff
@@ -20,12 +19,22 @@
  */
 public class SimpleCrac extends AbstractIdentifiable implements Crac {
     private List<Cnec> cnecs;
+    private List<Contingency> contingencies;
```
Comment (terriervik) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/SimpleCrac.java]: Why did you change the format?

File: data/crac-file/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Synchronizable.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (terriervik) [reply data/crac-file/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Synchronizable.java]: > add a method `desynchronize` ??

Why?

File: data/crac-file/crac-impl/src/test/java/com/farao_community/farao/data/crac_impl/threshold/AbsoluteFlowThresholdTest.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    @Test
+    public void isMinThresholdOvercome() throws Exception {
+        assertFalse(absoluteFlowThreshold.isMinThresholdOvercome(network, cnec));
+    }
+
+    @Test
+    public void isMaxThresholdOvercome() {
+    }
```
Comment (terriervik) [inline data/crac-file/crac-impl/src/test/java/com/farao_community/farao/data/crac_impl/threshold/AbsoluteFlowThresholdTest.java]: Please write something, or cancel it.

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/VoltageThreshold.java [hunk 13]
```diff
@@ -41,4 +45,24 @@ public double getMaxValue() {
...
+
+    @Override
+    public void synchronize(Network network, Cnec cnec) {
+
+    }
+
+    @Override
+    public void desynchronize() {
+
+    }
```
Comment (terriervik) [inline data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/VoltageThreshold.java]: Same

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/AbsoluteFlowThreshold.java (hunk 8 above)
Comment (TheMaskedTurtle) [reply data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/AbsoluteFlowThreshold.java]: This is actually the implementation, because there is no minimum threshold for AbsoluteFlowThreshold so it's normal this method always returns false.

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/RelativeFlowThreshold.java (hunk 9 above)
Comment (TheMaskedTurtle) [reply data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/RelativeFlowThreshold.java]: Same answer

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/VoltageThreshold.java (hunk 10 above)
Comment (TheMaskedTurtle) [reply data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/VoltageThreshold.java]: This is TODO, I replaced by NotImplementedException.

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/VoltageThreshold.java (hunk 13 above)
Comment (TheMaskedTurtle) [reply data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/threshold/VoltageThreshold.java]: I declared a default (empty) implementation in the abstract method so we don't have to re-implement empty methods in sub-classes.

File: data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/SimpleCrac.java (hunk 11 above)
Comment (TheMaskedTurtle) [reply data/crac-file/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/SimpleCrac.java]: It enables to create separately contingencies and cnecs
